packages/http/src/mocker/callback/callbacks.ts

```diff
@@ -0,0 +1,97 @@
+import { IHttpCallbackOperation, IHttpOperationRequest } from '@stoplight/types';
+import { resolveRuntimeExpressions } from '../../utils/runtimeExpression';
+import { IHttpRequest, IHttpResponse } from '../../types';
+import fetch  from 'node-fetch';
+import * as Option from 'fp-ts/lib/Option';
+import * as Either from 'fp-ts/lib/Either';
+import { map, reduce } from 'fp-ts/lib/Array';
+import * as Reader from 'fp-ts/lib/Reader';
+import * as TaskEither from 'fp-ts/lib/TaskEither';
+import { head } from 'fp-ts/lib/Array';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { generate as generateHttpParam } from '../generator/HttpParamGenerator';
+import { validateOutput } from '../../validator';
+import { parseResponse } from '../../utils/parseResponse';
+import withLogger from '../../withLogger';
+import { violationLogger } from '../../utils/logger';
+import { Logger } from 'pino';
+
+export function runCallback({ callback, request, response }: { callback: IHttpCallbackOperation, request: IHttpRequest, response: IHttpResponse }): Reader.Reader<Logger, TaskEither.TaskEither<Error, unknown>> {
+  return withLogger(logger => {
+    const { url, ...requestInit } = assembleRequest({ resource: callback, request, response });
+    const logViolation = violationLogger(logger);
+
+    logger.info({ name: 'CALLBACK' }, `${callback.callbackName}: Making request to ${url}...`);
+
+    return pipe(
+      TaskEither.tryCatch(() => fetch(url, requestInit), Either.toError),
+      TaskEither.chain(parseResponse),
+      TaskEither.map(element => {
+        logger.info({ name: 'CALLBACK' }, `${callback.callbackName}: Request finished`);
+
+        return pipe(
+          validateOutput({ resource: callback, element }),
+          Either.mapLeft(violations => {
+            pipe(violations, map(logViolation));
+            return violations;
+          }),
+          Either.alt(() => Either.right(element)),
+        );
+      }),
+      TaskEither.mapLeft(error => {
+        logger.error({ name: 'CALLBACK' }, `${callback.callbackName}: Request failed: ${error.message}`);
+        return error;
+      }),
```

Comment:
This flow here is a little bit weird; I think the reason why the code's been wrote in this way is that `chain` carries on the same error type — and so we can't move from `Error` to `NonEmptyArray<IPrismDiagnostic[]>`.

However, since we're only logging the error and not "handling" it for real, we can use a little trick to make this flow a little bit better. I've prepared a PR so you can take a look into and merge if if you agree:

https://github.com/stoplightio/prism/pull/772

In any case I'm going to look a little bit into the theory and see how you handle such situation properly, and then we can refactor in case that makes sense.

Reply:
#772 applied